components/batch-system/src/batch.rs

```diff
@@ -303,30 +289,43 @@ impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
                     batch.release_control(&self.router.control_box, len);
                 }
             }
-            if !batch.normals.is_empty() {
-                for (i, p) in batch.normals.iter_mut().enumerate() {
-                    let len = self.handler.handle_normal(p);
-                    batch.counters[i] += 1;
-                    if p.is_stopped() {
-                        reschedule_fsms.push((i, ReschedulePolicy::Remove));
-                    } else {
-                        if batch.counters[i] > 3 {
-                            hot_fsm_count += 1;
-                            // We should only reschedule a half of the hot regions, otherwise,
-                            // it's possible all the hot regions are fetched in a batch the
-                            // next time.
-                            if hot_fsm_count % 2 == 0 {
-                                reschedule_fsms.push((i, ReschedulePolicy::Schedule));
-                                continue;
-                            }
-                        }
-                        if let Some(l) = len {
-                            reschedule_fsms.push((i, ReschedulePolicy::Release(l)));
+
+            let mut hot_fsm_count = 0;
+            let mut fsm_cnt = 0;
+            while fsm_cnt < max_batch_size {
```

Comment:
The loop is hard to understand. How about divide it into two loops? The first one is looping existing fsms. The second one is looping those pendings in channel.

Reply:
Ok.